### ISSUE TITLE
Sharing resources between command_queues requires event objects

### ISSUE THREAD

**Issue by paulfradgley** [KhronosGroup/OpenCL-Docs#1114]
https://github.com/KhronosGroup/OpenCL-Docs/blob/00c79690be2454ccc3a9e7cfd348af5d24fa4869/api/appendix_a.asciidoc

> The application needs to implement appropriate synchronization across threads on the host processor to ensure that the changes to the state of a shared object (such as a command-queue object, memory object, program or kernel object) happen in the correct order (deemed correct by the application) when multiple command-queues in multiple threads are making changes to the state of a shared object.
>
> A command-queue can cache changes to the state of a memory object on the device associated with the command-queue. To synchronize changes to a memory object across command-queues, the application must do the following:
>
> In the command-queue that includes commands that modify the state of a memory object, the application must do the following:
> * Get appropriate event objects for commands that modify the state of the shared memory object.
> * Call the {clFlush} (or {clFinish}) API to issue any outstanding commands from this command-queue.
> 
> In the command-queue that wants to synchronize to the latest state of a memory object, commands queued by **the application must use the appropriate event objects** that represent commands that modify the state of the shared memory object as event objects to wait on. This is to ensure that commands that use this shared memory object complete in the previous command-queue before the memory objects are used by commands executing in this command-queue.
>
> The results of modifying a shared resource in one command-queue **while it is being used by another command-queue** are undefined.

The Spec states that the only way an application can ensure data synchronisation between command-queues is to use event objects.

This implies a behavioural difference between these two examples:

>clEnqueueWriteBuffer(queueA, memobj, CL_BLOCKING, 0, NULL, NULL);
>clEnqueueReadBuffer(queueB, memobj, CL_BLOCKING, 0, NULL, NULL);

and

>clEnqueueWriteBuffer(queueA, memobj, CL_BLOCKING, 0, NULL, &write_event);
>clEnqueueReadBuffer(queueB, memobj, CL_BLOCKING, 1, &write_event, NULL);

Isn't it sufficient to say that "modifying a shared resource in one command-queue while it is being used by another command-queue" is undefined?
What is the reason for strictly requiring event objects?
If two commands don't overlap, but also don't use event objects, what can be assumed about the contents of the shared resource?

**Comment by bashbaug** [comment]
Discussed in the April 16th teleconference.

-	The text in question is from an Appendix and is largely unchanged from the OpenCL 1.0 specification.  We suspect that it should have been updated when we updated the memory model in OpenCL 2.0.
-	We need to check that this case really is described formally in the memory model.  If it is, we can probably change the appendix to refer to the memory model vs. duplicating the information.

**Comment by bashbaug** [comment]
The appendix is basically describing three scenarios / requirements, and I believe all of them are already covered by the memory model, in the [Host-Side and Device-Side Commands](https://registry.khronos.org/OpenCL/specs/3.0-unified/html/OpenCL_API.html#_host_side_and_device_side_commands) section:

1. "A command-queue can cache changes to the state of a memory object on the device associated with the command-queue. To synchronize changes to a memory object across command-queues, the application must do the following: ..."

This scenario is covered by numbered items (2) and (3) in the memory model section.  Several other numbered items also apply and are not mentioned in the appendix.

2. "The results of modifying a shared resource in one command-queue while it is being used by another command-queue are undefined."

I think this scenario is covered by the defintion of a data race and "any such data race results in undefined behavior."

So, I think we can just remove these scenarios from the appendix, and reference the memory mode instead.  I will make a PR for review.
